### PR TITLE
Combine `SpamCheckerApi` with the more generic `ModuleApi`.

### PR DESCRIPTION
**Pull request by richvdh**
Lots of different module apis is not easy to maintain.

Rather than adding yet another `ModuleApi(hs, hs.get_auth_handler())` incantation, first add an `hs.get_module_api()` method and use it where possible.

### CODE REVIEW

File: synapse/events/spamcheck.py [hunk 1]
```diff
@@ -17,24 +17,25 @@
 import inspect
 from typing import Any, Dict, List, Optional, Tuple
 
-from synapse.spam_checker_api import RegistrationBehaviour, SpamCheckerApi
+from synapse.spam_checker_api import RegistrationBehaviour
 from synapse.types import Collection
 
 MYPY = False
 if MYPY:
+    import synapse.events
```
Comment (clokep) [inline synapse/events/spamcheck.py:25]: Why this change?

File: synapse/server.py [hunk 2]
```diff
@@ -656,6 +657,10 @@ def get_replication_streams(self) -> Dict[str, Stream]:
     def get_federation_ratelimiter(self) -> FederationRateLimiter:
         return FederationRateLimiter(self.clock, config=self.config.rc_federation)
 
+    @cache_in_self
```
Comment (clokep) [inline synapse/server.py:660]: I wonder if this should *not* be cached? Caching it means that different plugins could theoretically interfere with each other.

**Review by clokep — APPROVED** [review]
Seems reasonable! Left a couple of questions, but nothing big. I think the failing CI was from the tests you just fixed.

File: synapse/events/spamcheck.py (hunk 1 above)
Comment (richvdh) [reply synapse/events/spamcheck.py:25]: well, because `synapse.events` is referenced at line 42. I'm not really sure why mypy wasn't complaining before, but I think including it here is more correct.

File: synapse/server.py (hunk 2 above)
Comment (richvdh) [reply synapse/server.py:660]: well, they can probably do that anyway, if they so desire, since most of the functions are thin wrappers. :man_shrugging:

File: synapse/events/spamcheck.py (hunk 1 above)
Comment (clokep) [reply synapse/events/spamcheck.py:25]: Looks like mypy checks this file. 🤷

File: synapse/server.py (hunk 2 above)
Comment (clokep) [reply synapse/server.py:660]: True, probably not worth trying to protect against without a big effort.
